privacyidea/api/validate.py

```diff
@@ -158,6 +162,43 @@ def after_request(response):
     return response
 
 
+@validate_blueprint.route('/offlinerefill', methods=['POST'])
+@event("validate_offlinerefill", request, g)
+def offlinerefill():
+    """
+    This endpoint allows to fetch new offline OTP values for a token,
+    that is already offline.
+    According to the definition it will send the missing OTP values, so that
+    the client will have as much otp values as defined.
+
+    :param serial: The serial number of the token, that should be refilled.
+    :param refilltoken: The authorization token, that allows refilling.
+    :param pass: the last password (maybe password+OTP) entered by the user
+    :return:
+    """
+    refilltoken = None
+    otps = {}
+    serial = getParam(request.all_data, "serial", required)
```

Comment:
In case the serial has no matching token, the following response is generated:
```json
{
    "detail": null, 
    "id": 1, 
    "jsonrpc": "2.0", 
    "result": {
        "error": {
            "code": -500, 
            "message": "'NoneType' object has no attribute 'machine_list'"
        }, 
        "status": false
    }, 
    ...
}
```

This error is raised:
```
Traceback (most recent call last):
  File "/home/fred/privacyidea/privacyidea/venv/lib/python2.7/site-packages/flask/app.py", line 1817, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/fred/privacyidea/privacyidea/venv/lib/python2.7/site-packages/flask/app.py", line 1477, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/fred/privacyidea/privacyidea/venv/lib/python2.7/site-packages/flask/app.py", line 1381, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/fred/privacyidea/privacyidea/venv/lib/python2.7/site-packages/flask/app.py", line 1475, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/fred/privacyidea/privacyidea/venv/lib/python2.7/site-packages/flask/app.py", line 1461, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/event.py", line 60, in event_wrapper
    f_result = func(*args, **kwds)
  File "/home/fred/privacyidea/privacyidea/privacyidea/api/validate.py", line 184, in offlinerefill
    machine_defs = list_token_machines(serial)
  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/log.py", line 154, in log_wrapper
    return func(*args, **kwds)
  File "/home/fred/privacyidea/privacyidea/privacyidea/lib/machine.py", line 332, in list_token_machines
    for machine in db_token.machine_list:
AttributeError: 'NoneType' object has no attribute 'machine_list'
```
Maybe we should return a descriptive error message.

Reply:
Is this still a problem or is this outdated due to your rewrite?

Reply:
I forgot about this :) ce085ac now deals with error cases.